### DISCUSSION TITLE
DSN file saved from FreeRouting won't load again in FreeRouting

### DISCUSSION THREAD

**Discussion by lynchaj** [freerouting/freerouting#187] [Q&A]
Hi
I am using KiCAD 6.0.11 and exported a DSN file to FreeRouting 1.7.0.  I am using the latest OpenJRE on Win32 and the FreeRouting 1.7.0 .jar file.

FreeRouting opens the initial (components but no traces) DSN file and starts the autorouting, finishes and starts optimizing just fine.

However, when I stop FreeRouting and save the file as a DSN but then try to reload the DSN I get an empty circuit board with no components or traces.  This not what I would expect.  I think it should load the DSN file with components and traces just like I saved it.

Any ideas on what is going on?

Thanks, Andrew Lynch
[processor.Z80.zip](https://github.com/freerouting/freerouting/files/11301855/processor.Z80.zip)

**Comment by andrasfuchs** [comment]
Hi Andrew,

It might be just a typo, but you export your finished board after autorouting as a .SES file, right? You need a .SES file (not a .DSN) as a result, and import that back to KiCad.

The other thing I noticed is that your input file is rather big (~2 MBs). It caused problems for others as well, so I made a fix, but it's only available in [our snapshot build](https://github.com/freerouting/freerouting/releases/tag/SNAPSHOT). 
Could you re-try the routing with that?

Regards,
Andras

**Reply by lynchaj** [reply]
Hi, I downloaded the latest snapshot build and tried loading the .DSN file but it still shows a blank board outline with no components or traces.

Is it possible to load the empty .DSN exported from KiCAD (without traces) and then reimport the .SES into FreeRouting?

Thanks, Andrew Lynch

**Comment by lynchaj** [comment]
HiYes, I export a .SES file to reimport into KiCAD usually at the end of the optimization phase.  However, what I am doing is prior to importing into KiCAD I am occasionally interrupting optimization and saving the .DSN file (with the traces) so in case I lose power or my computer crashes that I don't lose everything and have to start over from scratch.
I will pull down the snapshot build and give that a try and report back. 
Thank you!  Andrew Lynch

    On Saturday, April 22, 2023 at 03:10:37 PM EDT, Andras Fuchs ***@***.***> wrote:  
 
 


Hi Andrew,

It might be just a typo, but you export your finished board after autorouting as a .SES file, right? You need a .SES file (not a .DSN) as a result, and import that back to KiCad.

The other thing I noticed is that your input file is rather big (~2 MBs). It caused problems for others as well, so I made a fix, but it's only available in our snapshot build.
Could you re-try the routing with that?

Regards,
Andras

**Comment by andrasfuchs** [comment]
Thank you for the clarification @lynchaj!

I can confirm that there is a bug in the DSN reader and potentially in the writer as well.

The following part causes the problem:
```
  (placement
    (component "Connector_IDC:IDC-Header_2x25_P2.54mm_Horizontal"
      (place 
        P1 108000.0 -223000.0 front 270
        (pin 1 (clearance_class Power))
        (pin 2 (clearance_class Power))

```

The reader can't make sense of the third line, because it doesn't expect to be a new line after the "place" command. 
I'll need to double check the DSN spec if this is allowed, and fix the issue with the reader/writer accordingly.

**Reply by andrasfuchs** [reply]
Alright, I posted a fix, it works here now.

![image](https://user-images.githubusercontent.com/910321/233934076-57a742b5-0644-4bec-a1f3-50686e75e474.png)

Could you test the most recent snapshot build, and let me know if it's ok?

**Reply by lynchaj** [reply]
Hi
I downloaded the latest snapshot build and tried again with my project .DSN file.  It does load much better and now I can see the components and *some* traces.  However, as you can see above a lot of traces are left out and results in incomplete nets.  The .DSN I saved has trace routing complete and was in the optimization stage so there shouldn't be any incomplete nets.

Maybe this is a different problem or another aspect of the same problem?

Thanks!  Andrew Lynch

**Comment by lynchaj** [comment]
Thank you!  Much appreciated!
Andrew Lynch
    On Monday, April 24, 2023 at 03:39:46 AM EDT, Andras Fuchs ***@***.***> wrote:  
 
 


Thank you for the clarification @lynchaj!

I can confirm that there is a bug in the DSN reader and potentially in the writer as well.

The following part causes the problem:
  (placement
    (component "Connector_IDC:IDC-Header_2x25_P2.54mm_Horizontal"
      (place 
        P1 108000.0 -223000.0 front 270
        (pin 1 (clearance_class Power))
        (pin 2 (clearance_class Power))


The reader can't make sense of the third line, because it doesn't expect to be a new line after the "place" command.
I'll need to double check the DSN spec if this is allowed, and fix the issue with the reader/writer accordingly.

**Comment by lynchaj** [comment]
HiOK, will do later today
Thank you!  Andrew Lynch
    On Monday, April 24, 2023 at 03:58:24 AM EDT, Andras Fuchs ***@***.***> wrote:  
 
 


Alright, I posted a fix, it works here now.



Could you test the most recent snapshot build, and let me know if it's ok?

**Comment by andrasfuchs** [comment]
I think @pierremolinaro is on the right track here: https://github.com/freerouting/freerouting/issues/182#issuecomment-1491957510

If you still have "{" and "~" characters in the names then they can cause problems.

What I can see as a clear improvement path for Freerouting is that we should display all warning and error messages after we try to load a DSN file and let the user know the reason for the (partially) failed loading process.
